wave_utils/include/wave/utils/config.hpp

```diff
@@ -0,0 +1,213 @@
+#ifndef __WAVE_UTILS_CONFIG_HPP__
+#define __WAVE_UTILS_CONFIG_HPP__
+
+#include <type_traits>
+
+#include <yaml-cpp/yaml.h>
+
+#include <opencv2/core/core.hpp>
+#include <opencv2/imgproc/imgproc.hpp>
+#include <opencv2/highgui/highgui.hpp>
+
+#include "wave/utils/math.hpp"
+#include "wave/utils/logging.hpp"
+#include "wave/utils/filesystem.hpp"
+
+
+namespace wave {
+
+enum ConfigDataType {
+    TYPE_NOT_SET = 0,
+    // PRIMITIVES
+    BOOL = 1,
+    INT = 2,
+    FLOAT = 3,
+    DOUBLE = 4,
+    STRING = 5,
+    // ARRAY
+    BOOL_ARRAY = 11,
+    INT_ARRAY = 12,
+    FLOAT_ARRAY = 13,
+    DOUBLE_ARRAY = 14,
+    STRING_ARRAY = 15,
+    // VECTOR
+    VEC2 = 21,
+    VEC3 = 22,
+    VEC4 = 23,
+    VECX = 24,
+    // MATRIX
+    MAT2 = 31,
+    MAT3 = 32,
+    MAT4 = 33,
+    MATX = 34,
+    CVMAT = 35
+};
+
+class ConfigParam {
+public:
+    enum ConfigDataType type;
+    std::string key;
+    bool optional;
+
+    bool *b;
```

Comment:
"ConfigParam a class template and using template specialization, without ever storing param.type in memory or checking it at runtime" 

That is a good point, but I chose the current `addParam()` implementation because I wanted to be able to inspect the logic for loading simliar different types in one shot (e.g. primitives, std::vectors, vectors, matrices). 

The `ConfigParam` itself can be considered the `union`, I think converting the list of pointers to a union does not add much to this specific class at the moment.